metricbeat/internal/metrics/memory/memory.go

```diff
@@ -0,0 +1,124 @@
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package memory
+
+import (
+	"github.com/pkg/errors"
+
+	"github.com/elastic/beats/v7/libbeat/common"
+	"github.com/elastic/beats/v7/metricbeat/internal/metrics"
+	"github.com/elastic/go-structform/gotype"
+)
+
+// Memory holds os-specifc memory usage data
+// The vast majority of these values are cross-platform
+// However, we're wrapping all them for the sake of safety, and for the more variable swap metrics
+type Memory struct {
+	Total metrics.OptUint `struct:"total,omitempty"`
+	Used  UsedMemStats    `struct:"used,omitempty"`
+
+	Free   metrics.OptUint `struct:"free,omitempty"`
+	Cached metrics.OptUint `struct:"cached,omitempty"`
+	// "Actual" values are, technically, a linux-only concept
+	// For better or worse we've expanded it to include "derived"
+	// Memory values on other platforms, which we should
+	// probably keep for the sake of backwards compatibility
+	// However, because the derived value varies from platform to platform,
+	// We may want to more precisely document what these mean.
+	Actual ActualMemoryMetrics `struct:"actual,omitempty"`
+
+	// Swap metrics
+	Swap SwapMetrics `struct:"swap,omitempty"`
+}
+
+// UsedMemStats wraps used.* memory metrics
+type UsedMemStats struct {
+	Pct   metrics.OptFloat `struct:"pct,omitempty"`
+	Bytes metrics.OptUint  `struct:"bytes,omitempty"`
+}
+
+// ActualMemoryMetrics wraps the actual.* memory metrics
+type ActualMemoryMetrics struct {
+	Free metrics.OptUint `struct:"free,omitempty"`
+	Used UsedMemStats    `struct:"used,omitempty"`
+}
+
+// SwapMetrics wraps swap.* memory metrics
+type SwapMetrics struct {
+	Total metrics.OptUint `struct:"total,omitempty"`
+	Used  UsedMemStats    `struct:"used,omitempty"`
+	Free  metrics.OptUint `struct:"free,omitempty"`
+}
+
+// Get returns platform-independent memory metrics.
+func Get(procfs string) (Memory, error) {
+	base, err := get(procfs)
+	if err != nil {
+		return Memory{}, errors.Wrap(err, "error getting system memory info")
+	}
+	base.fillPercentages()
+	return base, nil
+}
+
+// IsZero implements the zeroer interface for structform's folders
+func (used UsedMemStats) IsZero() bool {
+	return used.Pct.IsZero() && used.Bytes.IsZero()
+}
+
+// IsZero implements the zeroer interface for structform's folders
+func (swap SwapMetrics) IsZero() bool {
+	return swap.Free.IsZero() && swap.Used.IsZero() && swap.Total.IsZero()
+}
+
+func (base *Memory) fillPercentages() {
+	// Add percentages
+	// In theory, `Used` and `Total` are available everywhere, so assume values are good.
+	if base.Total.ValueOr(0) != 0 {
+		percUsed := float64(base.Used.Bytes.ValueOr(0)) / float64(base.Total.ValueOr(0))
```

Comment:
looks like we have a chance to divide by 0.

Reply:
That's what the `if` statement should be for, but I changed it to `ValueOr(1)` anyway.